indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsTask.java

```diff
@@ -218,7 +221,22 @@ public TaskStatus runTask(TaskToolbox toolbox) throws Exception
       // abandoned.
 
       toolbox.getTaskActionClient().submit(new SegmentNukeAction(new HashSet<>(unusedSegments)));
-      toolbox.getDataSegmentKiller().kill(unusedSegments);
+
+      // Fetch the load specs of all segments overlapping with the given interval
+      final Set<Map<String, Object>> usedSegmentLoadSpecs = toolbox
+          .getTaskActionClient()
+          .submit(new RetrieveUsedSegmentsAction(getDataSource(), getInterval(), null, Segments.INCLUDING_OVERSHADOWED))
+          .stream()
+          .map(DataSegment::getLoadSpec)
+          .collect(Collectors.toSet());
+
+      // Kill segments from the deep storage only if their load specs are not being used by any used segments
+      final List<DataSegment> segmentsToBeKilled = unusedSegments
+          .stream()
+          .filter(unusedSegment -> !usedSegmentLoadSpecs.contains(unusedSegment.getLoadSpec()))
+          .collect(Collectors.toList());
+
+      toolbox.getDataSegmentKiller().kill(segmentsToBeKilled);
```

Comment:
This action is very expensive action especially if the interval is eternity which is often the case with killUnusedSegments  task.

Reply:
This was already happening, this PR has just reduced the number of segments that are being killed by the `DataSegmentKiller`.

Reply:
How is that Kashif? Were we getting the list of used segments before too?

Reply:
Ah, sorry, I misinterpreted the comment to be just for the last line, rather than this whole block of code.